### PR TITLE
multiclient: assign client an unique id

### PR DESCRIPTION
**Pull request by markya0616**

### CODE REVIEW

File: multiclient/map.go [hunk 1]
```diff
@@ -24,65 +24,104 @@ import (
 )
 
 type Map struct {
-	m           map[string]*rpc.Client
-	newClientCh chan<- string
+	// the mapping form url to client
+	clientMap map[string]*client
+	// the mapping subscription id to url
+	idMap              map[uint64]string
+	subscripionCounter uint64
```
Comment (tailingchen) [inline multiclient/map.go]: s/subscripionCounter/clientCounter

File: multiclient/client.go [hunk 2]
```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
...
+			err := doSubscribe(ctx, id, rc, ch)
+			if err == nil {
+				return nil
 			}
 		}
+
 		// retry subscribe after retryPeriod
-		time.Sleep(retryPeriod)
-		log.Trace("Retry to subscribe new head", "url", url)
+		t := time.NewTicker(retryPeriod)
```
Comment (tailingchen) [inline multiclient/client.go]: Prefer to use time.Timer.
Declare timer outside for-loop.
```
retryTimer := time.NewTimer(0)
defer retryTimer.Stop()
for {
  ......
  retryTimer.Reset(retryPeriod)
  select {
  case <-retryTimer.C:
  case <-ctx.Done():
    return nil
  }
}
```

File: multiclient/client.go [hunk 3]
```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
...
-func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, url string, ch chan<- *Header) error {
+func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, id uint64, ch chan<- *Header) error {
+	logger := log.New("id", id)
 	defer wg.Done()
 
 	for {
-		rc := mc.rpcClientMap.Get(url)
-		if rc == nil {
-			log.Trace("EthClient has been removed", "url", url)
+		rc, existed := mc.rpcClientMap.GetById(id)
```
Comment (tailingchen) [inline multiclient/client.go]: How about return url?
`url` is more useful than `id` to logging.

File: multiclient/client.go [hunk 4]
```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
...
+		case <-ctx.Done():
+			t.Stop()
+			return nil
+		}
+		t.Stop()
+		logger.Trace("Retry to subscribe new head")
+	}
+}
+
+func doSubscribe(ctx context.Context, id uint64, rc *rpc.Client, ch chan<- *Header) error {
```
Comment (tailingchen) [inline multiclient/client.go]: Replace id with url.

File: multiclient/client.go [hunk 5]
```diff
@@ -657,46 +657,69 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
...
+	logger := log.New("id", id)
+	headerCh := make(chan *types.Header)
+	c := ethclient.NewClient(rc)
+	subCtx, cancel := context.WithCancel(ctx)
```
Comment (tailingchen) [inline multiclient/client.go:694]: Why we need to create cancel context here?
SubscribeNewHead seems to be canceled only by client.

File: multiclient/map.go (hunk 1 above)
Comment (markya0616) [reply multiclient/map.go]: rename to `idCounter`

File: multiclient/client.go (hunk 2 above)
Comment (markya0616) [reply multiclient/client.go]: OK

File: multiclient/client.go (hunk 3 above)
Comment (markya0616) [reply multiclient/client.go]: OK

File: multiclient/client.go (hunk 4 above)
Comment (markya0616) [reply multiclient/client.go]: OK, pass a logger with `id` and `url` in.

File: multiclient/client.go (hunk 5 above)
Comment (markya0616) [reply multiclient/client.go:694]: Just ensure the `SubscribeNewHead` call is closed.

File: multiclient/client.go [hunk 6]
```diff
@@ -657,46 +657,66 @@ func (mc *Client) SubscribeNewHead(ctx context.Context, ch chan<- *Header) (ethe
 	}), nil
 }
 
-func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, url string, ch chan<- *Header) error {
+func (mc *Client) subscribeNewHead(ctx context.Context, wg *sync.WaitGroup, id uint64, ch chan<- *Header) error {
+	logger := log.New("id", id)
```
Comment (tailingchen) [inline multiclient/client.go:661]: Remove it.

**Review by tailingchen — APPROVED** [review]

**Pull request merged**